File: frontend/src/apis/rating/GetUserRating.ts
import type { RatingValue } from '@/objects/rating/RatingValue'
import type { Username } from '@/objects/user/Username'
import type { APIWithSessionMessage } from '@/system/api/api-message'

/** 获取单个用户 rating 的内部 API；输入用户名作为 body，输出 rating 值。 */
export class GetUserRating implements APIWithSessionMessage<RatingValue> {
  declare readonly responseType?: RatingValue
  readonly method = 'POST'
  readonly apiPath: string
  private readonly username: Username

  constructor(username: Username) {
    this.username = username
    this.apiPath = 'internal/ratings/user'
  }

  body(): Username {
    return this.username
  }
}

File: backend/src/main/scala/domains/rating/api/GetUserRating.scala
package domains.rating.api

import cats.effect.IO
import domains.auth.api.InternalOnlyApi
import domains.rating.objects.RatingValue
import domains.rating.table.rating.RatingTable
import domains.user.objects.Username
import org.http4s.Method
import shared.api.ApiPath

import java.sql.Connection

/** 内部用户评分查询 API，供其他 domain 聚合用户展示信息时复用。 */
object GetUserRating extends InternalOnlyApi[Username, RatingValue]:

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/internal/ratings/user")

  /** 读取用户当前评分；没有评分状态时返回初始评分。 */
  override def plan(connection: Connection, username: Username): IO[RatingValue] =
    RatingTable.findUserRating(connection, username)
